research/cifar10/ditto/client.py

```diff
@@ -0,0 +1,188 @@
+import os
+from logging import INFO
+from pathlib import Path
+from typing import Dict, Optional, Sequence, Tuple
+
+import flwr as fl
+import torch
+import torch.nn as nn
+from flwr.common.logger import log
+from flwr.common.typing import Config
+from torch.nn.modules.loss import _Loss
+from torch.optim import Optimizer
+from torch.utils.data import DataLoader
+
+from fl4health.checkpointing.checkpointer import BestLossTorchCheckpointer
+from fl4health.checkpointing.client_module import ClientCheckpointModule
+from fl4health.clients.ditto_client import DittoClient
+from fl4health.utils.config import narrow_config_type
+from fl4health.utils.load_data import load_cifar10_data, load_cifar10_test_data
+from fl4health.utils.losses import LossMeterType
+from fl4health.utils.metrics import Accuracy, Metric
+from fl4health.utils.random import set_all_random_seeds
+from fl4health.utils.sampler import DirichletLabelBasedSampler
+from research.cifar10.model import ConvNet
+
+NUM_CLIENTS = 10
+
+
+class CifarDittoClient(DittoClient):
+    def __init__(
+        self,
+        data_path: Path,
+        metrics: Sequence[Metric],
+        device: torch.device,
+        client_number: int,
+        learning_rate: float,
+        lam: float,
+        heterogeneity_level: float,
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        checkpointer: Optional[ClientCheckpointModule] = None,
+    ) -> None:
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            checkpointer=checkpointer,
+            lam=lam,
+        )
+        self.client_number = client_number
+        self.heterogeneity_level = heterogeneity_level
+        self.learning_rate: float = learning_rate
+
+        assert 0 <= client_number < NUM_CLIENTS
+        log(INFO, f"Client Name: {self.client_name}, Client Number: {self.client_number}")
+
+    def get_data_loaders(self, config: Config) -> Tuple[DataLoader, DataLoader]:
+        batch_size = narrow_config_type(config, "batch_size", int)
+        n_clients = narrow_config_type(config, "n_clients", int)
+        # Set client-specific hash_key for sampler to ensure heterogneous data distribution among clients
+        sampler = DirichletLabelBasedSampler(
+            list(range(10)),
+            sample_percentage=1.0 / n_clients,
+            beta=self.heterogeneity_level,
+            hash_key=self.client_number,
+        )
+        # Set the same hash_key for the train_loader and val_loader to ensure the same data split
+        # of train and validation for all clients
+        train_loader, val_loader, _ = load_cifar10_data(
+            self.data_path,
+            batch_size,
+            validation_proportion=0.2,
+            sampler=sampler,
+            hash_key=100,
+        )
+        return train_loader, val_loader
+
+    def get_test_data_loader(self, config: Config) -> Optional[DataLoader]:
```

Comment:
For testing, we might want to reserve a test set that none of the clients see. It'll also be important to replicate the test set that they used in the benchmark. Here, we're using the sampler, but do they use just a uniform dataset, do they split the test set from the Dirichlet allocation etc.

Reply:
Right now the testing set is completely separate than training and evaluation as we are using cifar10 specific test split for it. What do you mean by reserving it separately? Regarding the paper approach let me recheck it again.

Reply:
Gotcha. I completely forgot that the CIFAR set we load here is different from that in the training/evaluation stage. Thanks for pointing that out.

Reply:
I checked that in their code, and they make test data follow train data. However their reported average is weighted average which in our case is not weighted. I think we need to change this to follow their setting as well.

Reply:
Yup that's fair. I don't think it's too hard to do weighted average. Though, depending on how we distribute the data etc. they could be pretty close to equivalent.

Reply:
Oh yeah indeed they are same :( which is not the best case, the point of heterogeneity is to also have various number of data points. I think I just want to use new implementation of heterogeneity.